### ISSUE TITLE
Add capability to have custom generators

### ISSUE THREAD

**Issue by WalternativE** [LambdaFactory/Fornax#35]
## Rationale

https://github.com/LambdaFactory/Fornax/issues/3 states the need to generate RSS feeds for Fornax sites. I can think about also adding atom feeds, sitemaps and other files that basically have a custom format and need information from the site model as well as the available posts.

Maintaining all these different generators within the main project could induce a lot of friction. Having the possibility to ad-hoc generate files based on the specific site configuration (as well as needs of the site creator) seems to be a good idea.

## Implementation

I built a working version with https://github.com/LambdaFactory/Fornax/pull/34. It works the same way the HTML version of posts are generated only that it would execute at the end of the site generation process and produce a single file per generator.

I specified the folder in which generators live as `_generators` in the root directory. A custom generator only has to be called `cenerate` and be of the type `CustomGenerator<'a>` where `'a` is the `SiteModel`.

The simplest possible custom generator could look like this

```fsharp
#load "../siteModel.fsx"

open SiteModel

let generate : CustomGenerator<SiteModel> =
    fun siteModel posts ->
        "Hello world", "/path/hello.txt"
```
The type `CustomGenerator<'a>` boils down to these types in the Fornax model

```fsharp

type FileContent = string
type VirtualFilePath = string
type CustomGeneratorResult = FileContent * VirtualFilePath

type CustomGenerator<'a> = 'a -> Post list -> CustomGeneratorResult
```

This would mean you only have to produce some string content (in any way you wish) and provide a path (like `/feed/feed.xml`) and you're good.

What do you all think about this? Would be glad to have your feedback on this to make this work 😊

**Comment by NatElkins** [comment]
@WalternativE I think this is great and have been thinking about doing something similar.

One thought I had was I wondered if we should make posts just another implementation of a custom generator.  Right now posts are "special" in that they get passed into every generator.  But what if you had other types of content across the site that you wanted to have access to?

What I was thinking is rather than `Post list` getting passed into every function, all the contents of the site could be available in each generator.  A sample implementation of that might look like this:

```fsharp
open System.Collections.Generic
open System.ComponentModel.Design

type SiteContents () =
    let container = new ServiceContainer()

    member __.Add(value:'a) = 
        let key = typeof<List<'a>>
        match container.GetService(key) with
        | :? List<'a> as result ->
            result.Add(value)
        | _ ->
            container.AddService(key, List<'a>([|value|]))
    
    member __.GetValues<'a> () : seq<'a> = 
        let key = typeof<List<'a>>
        let result = container.GetService(key)
        result :?> IEnumerable<'a>
    
    member this.TryGetValues<'a> () = 
        let key = typeof<List<'a>>
        if container.GetService(key) |> isNull then
            None
        else this.GetValues<'a>() |> Some

type Post = { A : int}
type OtherContent = { B : string}

[<EntryPoint>]
let main argv =
    // This is what would get passed into every function.
    // Assume it's been filled with content in advance, same
    // as the Post list is today.
    let siteContents = SiteContents()
    let post = { A = 1 }
    let post2 = { A = 2 }
    let otherContent = { B = "Test" }

    // You can see it's easy to fill with arbitrary stuff.
    siteContents.Add(post)
    siteContents.Add(post2)
    siteContents.Add(otherContent)

    // If you wanted a particular kind of content, here's
    // how you could get it in a typesafe way.
    let posts = siteContents.GetValues<Post>()
    let otherContents = siteContents.GetValues<OtherContent>()

    for i in posts do
        printfn "Post: %A" i
    for i in otherContents do
        printfn "Other Content %A" i
    0 // return an integer exit code

```

On the other hand, maybe this is just over-engineering things, since 95% of people just want a site that has some kind of main page, and a directory `/posts/` that has their blog posts in it.  But if you wanted to have a page that could display things from other pages, not just posts, here would be a way to do it.

**Comment by NatElkins** [comment]
And instead of these being a folder called `_generators`, everything could just be based around convention.  `posts.fsx` would look in the `posts` folder and apply to anything in there, `xyxContents.fsx` would look in `xyzContents` folder, etc.  If there's no matching folder, assume no contents and just execute the .fsx.

Obviously there's some edge cases not handled, I'm just spitballing here, but I feel like there's some potential.

**Comment by Krzysztof-Cieslak** [comment]
> One thought I had wondered was if we should make posts just another implementation of a custom generator.

If we'll go with this proposal in general I think then this is what we should do - wouldn't make sense to keep posts as something special. This would move project in direction of general static site generator rather than something obviously focused on just blogging.

**Comment by WalternativE** [comment]
Hi @NatElkins and @Krzysztof-Cieslak! Really liking the ideas. Especially the convention based generation of content. I'll have to think a bit about this and try some things out. As I've been ill for a week I'll have to do all the work which I haven't done previously first so this might take some days 🙇‍♂️

**Comment by Krzysztof-Cieslak** [comment]
OK, I'm totally convinced we should do this, including moving built-in post support to the custom generator. @WalternativE you have any time to take another shot at this or should I just use your PR and continue from it?

**Comment by Krzysztof-Cieslak** [comment]
I actually wonder if we should move even further - one set of scripts called `loaders` that will populate data in the proposed `SiteContents` state, and the other set called `generators` that will get `SiteContents` state and generate stuff.

**Comment by Krzysztof-Cieslak** [comment]
Or maybe rather `generators` should provide to functions - one for loading data that will be executed at the beginning and other for generating content that will be executed after every generator loaded content

**Comment by Krzysztof-Cieslak** [comment]
So the one thing I can't figure out is defining when the given generator is run. Let's assume we go with `loaders` and `generators` design. All `loaders` are run at the fornax startup to fill the content. Then we have 2 generators - one for generating posts, and one for RSS. The generator for posts needs to run multiple times - once per every *.md file (or Post object in collection?). The RSS generator needs run once - it's not connected to any file on the hard drive but just takes context and generate some content. 

So we need to have some way of saying "this generator should run for every Post" or "this generator should run for every XML file we've loaded" or "this generator should only run once"

**Comment by Krzysztof-Cieslak** [comment]
I've started implementation on https://github.com/LambdaFactory/Fornax/tree/new_design. The easiest way to understand what's going on from the design point of view is by looking at the `samples` folder. 
Currently, I've managed to get to work concept of `loaders` - scripts that will be executed before the generation to populate `SiteContents` (as defined by Nat above).
Next step is working on `generators` (and figuring out a good way of configuring them)

**Comment by NatElkins** [comment]
@Krzysztof-Cieslak Any chance you can open a draft PR?  I find it easier to follow along that way.

**Comment by Krzysztof-Cieslak** [comment]
https://github.com/LambdaFactory/Fornax/pull/40

**Comment by WalternativE** [comment]
Apologies for being so unresponsive. Still swamped with worked from a couple of sources. This might hold on for a while until things settle again. Thanks for your work on the redesign 🙇‍♂️

**Comment by Krzysztof-Cieslak** [comment]
Implemented in #40. Thanks for this discussion, it gave me lot of good (i hope) ideas.
